test/agent.js

```diff
@@ -40,6 +41,82 @@ tap.test('Agent', { skip: SKIP }, t => {
     t.throw(() => new Agent().get(''), InvalidArgumentError)
   })
 
+  t.test('Agent close and destroy', t => {
+    t.plan(2)
+
+    const wanted = 'payload'
+
+    const createServer = () => http.createServer((req, res) => {
+      res.setHeader('Content-Type', 'text/plain')
+      res.end(wanted)
+    })
+
+    const promisifyServerListen = server => new Promise((resolve, reject) => {
+      server.listen(0, err => {
+        if (err) {
+          reject(err)
+        } else {
+          resolve()
+        }
+      })
+    })
+
+    const _request = async (t, agent, server) => {
+      try {
+        const { statusCode, headers, body } = await request(`http://localhost:${server.address().port}`, { agent })
+        t.strictEqual(statusCode, 200, 'status code should be 200')
+        t.strictEqual(headers['content-type'], 'text/plain', 'header should be text/plain')
+        const bufs = []
+        for await (const buf of body) {
+          bufs.push(buf)
+        }
+        t.strictEqual(wanted, Buffer.concat(bufs).toString('utf8'), 'wanted data should be stream by body')
+      } catch (err) {
+        t.fail(err)
+      }
+    }
+
+    t.test('closes all origins', async t => {
+      t.plan(15)
+      const N = 3
+
+      const servers = Array.from(Array(N)).map(() => createServer())
+
+      const agent = new Agent()
+
+      await Promise.all(servers.map(server => promisifyServerListen(server)))
+      await Promise.all(servers.map(server => _request(t, agent, server)))
+      for (const server of servers) {
+        t.ok(server.listening, 'server should be listening')
+      }
+      await agent.close()
+      for (const ref of agent[kAgentCache].values()) {
+        const pool = ref.deref()
+        t.ok(pool.closed, 'pools should be closed')
+      }
+    })
+
+    t.test('destroys all origins', async t => {
+      t.plan(15)
+      const N = 3
+
+      const servers = Array.from(Array(N)).map(() => createServer())
+
+      const agent = new Agent()
+
+      await Promise.all(servers.map(server => promisifyServerListen(server)))
+      await Promise.all(servers.map(server => _request(t, agent, server)))
+      for (const server of servers) {
+        t.ok(server.listening, 'server should be listening')
+      }
+      await agent.destroy()
+      for (const ref of agent[kAgentCache].values()) {
+        const pool = ref.deref()
+        t.ok(pool.destroyed, 'pools should be destroyed')
```

Comment:
You should be able to test for this through disconnect events? No need for internal symbol.

Reply:
The internal symbol is required because it is the only way to reference the pools inside of the agent. Unless you're expecting the agent itself to emit some disconnect event?

Reply:
But yes I can make use of the pool disconnect events for the test - will try that approach now.

Reply:
Maybe agent should also emit disconnect?

Reply:
Yeah I could try to forward that along - seems partially related to this https://github.com/nodejs/undici/issues/529 too